File: 20250509.py
from collections import deque, defaultdict


class Solution:
    def __init__(self):
        self.m = 0
        self.n = 0
        self.vis = []
        self.dx = [0, 0, 1, -1]
        self.dy = [1, -1, 0, 0]

    def numIslands(self, grid):
        self.m, self.n = len(grid), len(grid[0])
        self.vis = [[False] * self.n for _ in range(self.m)]
        ret = 0

        for i in range(self.m):
            for j in range(self.n):
                if not self.vis[i][j] and grid[i][j] == "1":
                    ret += 1
                    self.bfs(grid, i, j)

        return ret

    def bfs(self, grid, i, j):
        q = deque()
        q.append((i, j))
        self.vis[i][j] = True

        while q:
            a, b = q.popleft()
            for k in range(4):
                x = a + self.dx[k]
                y = b + self.dy[k]
                if (
                    0 <= x < self.m
                    and 0 <= y < self.n
                    and not self.vis[x][y]
                    and grid[x][y] == "1"
                ):
                    q.append((x, y))
                    self.vis[x][y] = True

    def nearestExit(self, maze, entrance):
        m, n = len(maze), len(maze[0])
        vis = [[False for _ in range(n)] for _ in range(m)]
        dx = [0, 0, 1, -1]
        dy = [1, -1, 0, 0]

        q = deque()
        q.append((entrance[0], entrance[1]))
        vis[entrance[0]][entrance[1]] = True
        steps = 0

        while q:
            steps += 1

            for _ in range(q):
                a, b = q.popleft()
                for k in range(4):
                    x, y = a + dx[k], b + dy[k]
                    if (
                        0 <= x < m
                        and 0 <= y < n
                        and not vis[x][y]
                        and maze[x][y] == "."
                    ):
                        if x == 0 or x == m - 1 or y == 0 or y == n - 1:
                            return steps
                        q.append((x, y))
                        vis[x][y] = True
        return -1

    def canFinish(self, n, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        edges = defaultdict(list)
        in_degree = [0] * n

        for a, b in prerequisites:
            edges[b].append(a)
            in_degree[a] += 1

        q = deque([i for i in range(n) if in_degree[i] == 0])

        while q:
            t = q.popleft()
            for a in edges[t]:
                in_degree[a] -= 1
                if in_degree[a] == 0:
                    q.append(a)

        return all(degree == 0 for degree in in_degree)

    def solve(self, board):
        self.m, self.n = len(board), len(board[0])
        self.vis = [[False] * self.n for _ in range(self.m)]
        self.dx = [0, 0, 1, -1]
        self.dy = [1, -1, 0, 0]

        for i in range(self.m):
            if board[i][0] == "O":
                self.bfs(board, i, 0)
            if board[i][self.n - 1] == "O":
                self.bfs(board, i, self.n - 1)

        for i in range(self.n):
            if board[0][i] == "O":
                self.bfs(board, 0, i)
            if board[self.m - 1][i] == "O":
                self.bfs(board, self.m - 1, i)

        for i in range(self.m):
            for j in range(self.n):
                if board[i][j] == "O":
                    board[i][j] = "X"
                elif board[i][j] == ".":
                    board[i][j] = "O"

    def bfs(self, board, i, j):
        q = deque()
        q.append((i, j))
        board[i][j] = "."
        self.vis[i][j] = True

        while q:
            a, b = q.popleft()
            for k in range(4):
                x, y = a + self.dx[k], b + self.dy[k]
                if (
                    0 <= x < self.m
                    and 0 <= y < self.n
                    and not self.vis[x][y]
                    and board[x][y] == "O"
                ):
                    q.append((x, y))
                    board[x][y] = "."
                    self.vis[x][y] = True

File: 20250509.cc
#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int m, n;
    bool vis[301][301] = {0};
    int numIslands(vector<vector<char>> &grid)
    {
        m = grid.size(), n = grid[0].size();
        int ret = 0;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
            {
                if (!vis[i][j] && grid[i][j] == '1')
                {
                    ret++;
                    BFS(grid, i, j);
                }
            }

        return ret;
    }

    int dx[4] = {0, 0, 1, -1};
    int dy[4] = {1, -1, 0, 0};
    void BFS(vector<vector<char>> &grid, int i, int j)
    {
        queue<pair<int, int>> q;
        q.push({i, j});
        vis[i][j] = true;

        while (q.size())
        {
            int sz = q.size();
            while (sz--)
            {
                auto [a, b] = q.front();
                q.pop();
                for (int k = 0; k < 4; k++)
                {
                    int x = a + dx[k], y = b + dy[k];
                    if (x >= 0 && x < m && y >= 0 && y < n && !vis[x][y] && grid[x][y] == '1')
                    {
                        q.push({x, y});
                        vis[x][y] = true;
                    }
                }
            }
        }
    }

    int m, n;
    bool vis[101][101] = {0};
    int dx[4] = {0, 0, 1, -1};
    int dy[4] = {1, -1, 0, 0};
    int nearestExit(vector<vector<char>> &maze, vector<int> &entrance)
    {
        m = maze.size(), n = maze[0].size();

        int ret = 0;
        queue<pair<int, int>> q;
        q.push({entrance[0], entrance[1]});
        vis[entrance[0]][entrance[1]] = true;

        while (q.size())
        {
            ret++;
            int sz = q.size();
            while (sz--)
            {
                auto [a, b] = q.front();
                q.pop();
                for (int k = 0; k < 4; k++)
                {
                    int x = a + dx[k], y = b + dy[k];
                    if (x >= 0 && x < m && y >= 0 && y < n && !vis[x][y] && maze[x][y] == '.')
                    {
                        if (x == 0 || x == m - 1 || y == 0 || y == n - 1)
                            return ret;
                        q.push({x, y});
                        vis[x][y] = true;
                    }
                }
            }
        }

        return -1;
    }

    bool canFinish(int n, vector<vector<int>> &prerequisites)
    {
        unordered_map<int, vector<int>> edges;
        vector<int> in(n);

        for (auto &e : prerequisites)
        {
            int a = e[0], b = e[1];
            edges[b].push_back(a);
            in[a]++;
        }

        queue<int> q;
        for (int i = 0; i < n; i++)
            if (in[i] == 0)
                q.push(i);

        while (q.size())
        {
            int t = q.front();
            q.pop();
            for (auto a : edges[t])
                if (--in[a] == 0)
                    q.push(a);
        }

        for (int i = 0; i < n; i++)
            if (in[i])
                return false;
        return true;
    }

    int dx[4] = {0, 0, 1, -1};
    int dy[4] = {1, -1, 0, 0};
    int m, n;
    bool vis[201][201] = {0};
    void solve(vector<vector<char>> &board)
    {
        m = board.size(), n = board[0].size();

        for (int i = 0; i < m; i++)
        {
            if (board[i][0] == 'O')
                BFS(board, i, 0);
            if (board[i][n - 1] == 'O')
                BFS(board, i, n - 1);
        }

        for (int i = 0; i < n; i++)
        {
            if (board[0][i] == 'O')
                BFS(board, 0, i);
            if (board[m - 1][i] == 'O')
                BFS(board, m - 1, i);
        }

        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
            {
                if (board[i][j] == 'O')
                    board[i][j] = 'X';
                else if (board[i][j] == '.')
                    board[i][j] = 'O';
            }
    }

    void BFS(vector<vector<char>> &board, int i, int j)
    {
        queue<pair<int, int>> q;
        board[i][j] = '.';
        q.push({i, j});
        vis[i][j] = true;

        while (q.size())
        {
            auto [a, b] = q.front();
            q.pop();
            for (int k = 0; k < 4; k++)
            {
                int x = a + dx[k], y = b + dy[k];
                if (x >= 0 && x < m && y >= 0 && y < n && !vis[x][y] && board[x][y] == 'O')
                {
                    q.push({x, y});
                    board[x][y] = '.';
                    vis[x][y] = true;
                }
            }
        }
    }

    int ladderLength(string beginWord, string endWord, vector<string> &wordList)
    {
        unordered_set<string> hash(wordList.begin(), wordList.end());
        // unordered_map<string, int> in;

        if (!hash.count(endWord))
            return 0;

        int ret = 1;

        queue<string> q;
        q.push(beginWord);

        while (q.size())
        {
            ret++;
            int sz = q.size();
            while (sz--)
            {
                auto t = q.front();
                q.pop();
                for (int i = 0; i < t.size(); i++)
                {
                    string tmp = t;
                    for (char ch = 'a'; ch <= 'z'; ch++)
                    {
                        tmp[i] = ch;
                        if (hash.count(tmp))
                        {
                            if (tmp == endWord)
                                return ret;
                            q.push(tmp);
                        }
                    }
                }
            }
        }

        return 0;
    }

    int m, n;
    int dx[4] = {0, 0, 1, -1};
    int dy[4] = {1, -1, 0, 0};
    bool vis[51][51] = {0};
    int maxAreaOfIsland(vector<vector<int>> &grid)
    {
        m = grid.size(), n = grid[0].size();

        int ret = 0;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                if (grid[i][j] == 1)
                    ret = max(ret, bfs(grid, i, j));

        return ret;
    }

    int bfs(vector<vector<int>> &grid, int i, int j)
    {
        queue<pair<int, int>> q;
        q.push({i, j});
        vis[i][j] = true;
        int ret = 1;

        while (q.size())
        {
            auto [a, b] = q.front();
            q.pop();
            for (int k = 0; k < 4; k++)
            {
                int x = a + dx[k], y = b + dy[k];
                if (0 <= x < m && 0 <= y < n && !vis[x][y] && grid[x][y] == 1)
                {
                    ret++;
                    q.push({x, y});
                    vis[x][y] = true;
                }
            }
        }

        return ret;
    }

    string removeDuplicates(string s)
    {
        string ret;
        for (auto ch : s)
        {
            if (ret.size() && ret.back() == ch)
                ret.pop_back();
            else
                ret += ch;
        }
        return ret;
    }

    int ladderLength(string beginWord, string endWord, vector<string> &wordList)
    {
        unordered_set<string> hash(wordList.begin(), wordList.end());
        unordered_set<string> vis;

        if (!hash.count(endWord))
            return 0;

        int ret = 1;
        queue<string> q;
        q.push(beginWord);
        vis.insert(beginWord);

        while (q.size())
        {
            int sz = q.size();
            while (sz--)
            {
                auto t = q.front();
                q.pop();

                for (int i = 0; i < t.size(); i++)
                {
                    string tmp = t;
                    for (char ch = 'a'; ch <= 'z'; ch++)
                    {
                        tmp[i] = ch;
                        if (hash.count(tmp) && !vis.count(tmp))
                        {
                            if (tmp == endWord)
                                return ret;
                            ret++;
                            q.push(tmp);
                            vis.insert(tmp);
                        }
                    }
                }
            }
        }

        return 0;
    }

    bool canFinish(int n, vector<vector<int>> &prerequisites)
    {
        unordered_map<int, vector<int>> edges;
        vector<int> in(n);

        for (auto &e : prerequisites)
        {
            int a = e[0], b = e[1];
            edges[b].push_back(a);
            in[a]++;
        }

        queue<int> q;

        for (int i = 0; i < n; i++)
            if (in[i] == 0)
                q.push(i);

        while (q.size())
        {
            auto t = q.front();
            q.pop();

            for (auto a : edges[t])
                if (--in[a] == 0)
                    q.push(a);
        }

        for (int i = 0; i < n; i++)
            if (in[i])
                return false;
        return true;
    }
};